src/daft-plan/src/optimization/rules/push_down_projection.rs

```diff
@@ -0,0 +1,398 @@
+    ops::{Aggregate, Project, Source},
+    LogicalPlan,
+};
+
+use super::{ApplyOrder, OptimizerRule};
+
+#[derive(Default)]
+pub struct PushDownProjection {}
+
+impl PushDownProjection {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn try_optimize_project(
+        &self,
+        projection: &Project,
+        plan: &LogicalPlan,
+    ) -> DaftResult<Option<Arc<LogicalPlan>>> {
+        let upstream_plan = projection.input.clone();
+        let upstream_schema = upstream_plan.schema();
+
+        // First, drop this projection if it is a no-op.
+        let projection_is_noop = {
+            let maybe_column_names = projection
+                .projection
+                .iter()
+                .map(|e| match e {
+                    Expr::Column(colname) => Some(colname.clone()),
+                    _ => None,
+                })
+                .collect::<Option<Vec<_>>>();
+
+            let all_columns_equal = maybe_column_names.map_or(false, |column_names| {
+                column_names
+                    .iter()
+                    .zip(upstream_schema.names().iter())
+                    .map(|(scol, pcol)| scol.as_ref() == pcol)
+                    .all(identity)
+            });
+
+            let no_unused_columns = upstream_schema.names().len() == projection.projection.len();
+
+            all_columns_equal && no_unused_columns
+        };
+        if projection_is_noop {
+            // Projection discarded but new root node has not been looked at;
+            // look at the new root node.
+            let new_plan = self.try_optimize(&upstream_plan)?.unwrap_or(upstream_plan);
+            return Ok(Some(new_plan));
+        }
+
+        match upstream_plan.as_ref() {
+            LogicalPlan::Source(source) => {
+                // Prune unnecessary columns directly from the source.
+                let [required_columns] = &plan.required_columns()[..] else {
+                    panic!()
+                };
+                if required_columns.len() < upstream_schema.names().len() {
+                    let pruned_upstream_schema = upstream_schema
+                        .fields
+                        .iter()
+                        .filter_map(|(name, field)| {
+                            required_columns.contains(name).then(|| field.clone())
+                        })
+                        .collect::<Vec<_>>();
+                    let schema = Schema::new(pruned_upstream_schema)?;
+                    let new_source: LogicalPlan = Source::new(
+                        schema.into(),
+                        source.source_info.clone(),
+                        source.partition_spec.clone(),
+                    )
+                    .into();
+
+                    let new_plan = plan.with_new_children(&[new_source.into()]);
+                    // Retry optimization now that the upstream node is different.
+                    let new_plan = self.try_optimize(&new_plan)?.unwrap_or(new_plan);
```

Comment:
Hmm is this necessary in case the projection is now a no-op (if it's a simple column selection projection) and we therefore will want to drop it?